File: libavogadro/src/python/camera.cpp
// Last update: timvdm 18 June 2009
#include <boost/python.hpp>

#include <avogadro/camera.h>
#include <avogadro/glwidget.h>

using namespace boost::python;
using namespace Avogadro;

void export_Camera()
{

  const Eigen::Transform3d& (Camera::*modelview_ptr)() const = &Camera::modelview;
  Eigen::Vector3d (Camera::*unProject_ptr1)(const Eigen::Vector3d&) const = &Camera::unProject;
  Eigen::Vector3d (Camera::*unProject_ptr2)(const QPoint&, const Eigen::Vector3d&) const = &Camera::unProject;
  Eigen::Vector3d (Camera::*unProject_ptr3)(const QPoint&) const = &Camera::unProject;
  
  class_<Avogadro::Camera, boost::noncopyable>("Camera", "Representation of the view")
    //
    // constructors
    //
    .def(init<const GLWidget*>())
    .def(init<const GLWidget*, double>())
    .def(init<const Camera*>())
    
    //
    // read/write properties
    //
    .add_property("angleOfViewY", 
        &Camera::angleOfViewY, 
        &Camera::setAngleOfViewY,
        "The vertical viewing angle in degrees.")

    .add_property("modelview", 
        make_function(modelview_ptr, return_value_policy<return_by_value>()),
        &Camera::setModelview, 
        "4x4 \"modelview\" matrix representing the camera orientation and position.")

    //
    // read-only properties
    //
    .add_property("parent", 
        make_function(&Camera::parent, return_value_policy<reference_existing_object>()),
        "The parent GLWidget.")

    .add_property("backTransformedXAxis", 
        &Camera::backTransformedXAxis,
        "Returns a unit vector pointing toward the right, expressed in the "
        "scene's coordinate system. This is simply the "
        "unit vector that is mapped to (1,0,0) by the camera rotation.")

    .add_property("backTransformedYAxis", 
        &Camera::backTransformedYAxis,
        "Returns a unit vector pointing upward, expressed in the "
        "scene's coordinate system. This is simply the "
        "unit vector that is mapped to (0,1,0) by the camera rotation.")

    .add_property("backTransformedZAxis", 
        &Camera::backTransformedZAxis,
        "Returns a unit vector pointing toward the camera, expressed in the "
        "scene's coordinate system. This is simply the "
        "unit vector that is mapped to (0,0,1) by the camera rotation.")

    .add_property("transformedXAxis", 
        &Camera::transformedXAxis,
        "Returns a unit vector pointing in the x direction, expressed in the "
        "space coordinate system.")

    .add_property("transformedYAxis", 
        &Camera::transformedYAxis,
        "Returns a unit vector pointing in the y direction, expressed in the "
        "space coordinate system.")

    .add_property("transformedZAxis", 
        &Camera::transformedZAxis,
        "Returns a unit vector pointing in the z direction, expressed in the "
        "space coordinate system.")

    // real functions 
    .def("applyPerspective", 
        &Camera::applyPerspective,
        "Calls gluPerspective() with parameters automatically chosen "
        "for rendering the GLWidget's molecule with this camera. Should be called "
        "only in GL_PROJECTION matrix mode.")

    .def("applyModelview", 
        &Camera::applyModelview,
        "Calls glMultMatrix() with the camera's \"modelview\" matrix. Should be called "
        "only in GL_MODELVIEW matrix mode.")

    .def("initializeViewPoint", 
        &Camera::initializeViewPoint,
        "Sets up the camera so that it gives a nice view of the molecule loaded in the "
        "parent GLWidget. Typically you would call this method right after loading a molecule.")

    .def("distance", 
        &Camera::distance, 
        "Returns the distance between @a point and the camera.")

    .def("translate", 
        &Camera::translate, 
        "Multiply the camera's \"modelview\" matrix on the right by the translation of the given "
        "vector. As the translation is applied on the right, the vector is understood in "
        "the molecule's coordinate system. Use this method if you want to give the impression "
        "that the molecule is moving while the camera remains fixed. This is the equivalent "
        "of the OpenGL function glTranslate().")

    .def("pretranslate", 
        &Camera::pretranslate,
        "Multiply the camera's \"modelview\" matrix on the left by the translation of given "
        "vector. Because the translation is applied on the left, the vector is understood in "
        "the coordinate system obtained by applying the camera's matrix to the molecule's "
        "coordinate system. Use this method if you want to give the impression that the camera "
        "is moving while the molecule remains fixed.")
    
    .def("rotate", 
        &Camera::rotate, 
        "Multiply the camera's \"modelview\" matrix on the right by the rotation of the given "
        "angle and axis. As the rotation is applied on the right, the axis vector is "
        "understood in the molecule's coordinate system. Use this method if you want to give "
        "the impression that the molecule is rotating while the camera remains fixed. This is the "
        "equivalent of the OpenGL function glRotate(), except that here the angle is expressed "
        "in radians, not in degrees. After the rotation is multiplied, a normalization is "
        "performed to ensure that the camera matrix remains sane.")

    .def("prerotate", 
        &Camera::prerotate,
        "Multiply the camera's \"modelview\" matrix on the left by the rotation of the given "
        "angle and axis. Because the rotation is applied on the left, the axis vector is "
        "understood in the the coordinate system obtained by applying the camera's matrix to "
        "the molecule's coordinate system. Use this method if you want to give "
        "the impression that the camera is rotating while the molecule remains fixed. "
        "After the rotation is multiplied, a normalization is performed to ensure that the "
        "camera matrix remains sane.")

    .def("unProjectWithZ", 
        unProject_ptr1, 
        "Performs an unprojection from window coordinates to space coordinates.")

    .def("unProject", 
        unProject_ptr2,
        "Performs an unprojection from window coordinates to space coordinates, "
        "into the plane passing through a given reference point and parallel to the screen. "
        "Thus the returned vector is a point of that plane. The rationale is that "
        "when unprojecting 2D window coords to 3D space coords, there are a priori "
        "infinitely many solutions, and one has to be choose. This is equivalent to "
        "choosing a plane parallel to the screen.")

    .def("unProject", 
        unProject_ptr3, 
        "Performs an unprojection from window coordinates to space coordinates, "
        "into the plane passing through the molecule's center and parallel to the screen. "
        "Thus the returned vector is a point belonging to that plane.")

    .def("project", 
        &Camera::project, 
        "Performs a projection from space coordinates to window coordinates.")

    .def("normalize", 
        &Camera::normalize,
        "The linear component (ie the 3x3 topleft block) of the camera matrix must "
        "always be a rotation. But after several hundreds of operations on it, "
        "it can drift farther and farther away from being a rotation. This method "
        "normalizes the camera matrix so that the linear component is guaranteed to be "
        "a rotation. Concretely, it performs a Gram-Schmidt orthonormalization to "
        "transform the linear component into a nearby rotation. "
        "The bottom row must always have entries 0, 0, 0, 1. This function overwrites "
        "the bottom row with these values.")
    ;

}

File: libavogadro/src/python/unittest/camera.py
from PyQt4.Qt import *
from numpy import *
import Avogadro
import sys
import unittest


class TestCamera(unittest.TestCase):
  def setUp(self):
    # create the GLWidget and load the default engines
    self.glwidget = Avogadro.GLWidget()
    self.glwidget.loadDefaultEngines()
    
    self.molecule = Avogadro.molecules.addMolecule()
    self.molecule.addAtom()
    self.glwidget.molecule = self.molecule
    self.assertNotEqual(self.glwidget.camera, None)

  def tearDown(self):
    # create the GLWidget and load the default engines
    None
    

  def test_parent(self):
    self.assertNotEqual(self.glwidget.camera.parent, None)
  
  def test_angleOfViewY(self):
    self.assert_(self.glwidget.camera.angleOfViewY)
    self.glwidget.camera.angleOfViewY = 40.0
    self.assertEqual(self.glwidget.camera.angleOfViewY, 40.0)

  def test_modelview(self):
    self.glwidget.camera.modelview
    m = self.glwidget.camera.modelview
    self.glwidget.camera.modelview = m
  
  def test_various(self):
    self.glwidget.camera.applyPerspective()
    self.glwidget.camera.applyModelview()
    self.glwidget.camera.initializeViewPoint()
    dist = self.glwidget.camera.distance(array([0., 0., 0.]))
    self.glwidget.camera.translate(array([0., 0., 0.]))
    self.glwidget.camera.pretranslate(array([0., 0., 0.]))
    self.glwidget.camera.rotate(3.14, array([0., 0., 0.]))
    self.glwidget.camera.prerotate(3.14, array([0., 0., 0.]))
    self.glwidget.camera.normalize()
  
  def test_axes(self):
    self.glwidget.camera.transformedXAxis
    self.glwidget.camera.transformedYAxis
    self.glwidget.camera.transformedZAxis
    self.glwidget.camera.backTransformedXAxis
    self.glwidget.camera.backTransformedYAxis
    self.glwidget.camera.backTransformedZAxis
  
  def test_project(self):
    
    point = QPoint(10,20)
    self.assertEqual(len(self.glwidget.camera.unProject(point)), 3)
    self.assertEqual(len(self.glwidget.camera.unProject(point, array([1., 0., 0.]))), 3)
    # added to fix name conflict WithZ
    self.assertEqual(len(self.glwidget.camera.unProjectWithZ(array([1., 2., 0.]))), 3)

    self.assertEqual(len(self.glwidget.camera.project(array([1., 2., 3.]))), 3)






if __name__ == "__main__":
  # create a new application
  # (must be done before creating a GLWidget)
  app = QApplication(sys.argv)
  unittest.main()
  sys.exit(app.exec_())
